Add: Happy-eyeballs (try IPv4 and IPv6 connections simultaneously)

SlySven opened the pull request:
#### Brief overview of PR changes/additions
Overhauls the profile connection process to identify whether IPv4 and/or IPv6 protocol addresses are available for the entered details and if BOTH are to then try to connect with both at the same time (with a slight preference for IPv6 {by starting it first}). The first one to connect (including securely and/or via a proxy) wins out and terminates the process using the other address protocol. This technique is known as [Happy Eyeballs](https://en.wikipedia.org/wiki/Happy_Eyeballs).

#### Motivation for adding to Mudlet
To improve the end-user's experience if they are fortunate to be on a "dual-stack" (both protocols available) network.

#### Other info (issues closed, discussion etc)
The on-screen messages have been massaged somewhat to inform the user about the nature of the connection(s) being made. ALL the addresses that have been found for the Server are listed instead of just the first one {this includes a "numerus" translatable string which will need processing by yours truly!} Should a raw IP address be provided instead of a "URL" then the result of the "reverse" lookup (trying to get a host name from the IP address) is displayed instead of just repeating the address - if anything *is* found - if not that is also noted instead.

One other effect of Mudlet being given a raw IP address is that those are not normally included in an SSL/TLS certificate - so rendering a secure connection impossible in that case - this was already noted for an IPv4 address but NOT for an IPv6 one. This PR corrects that to also reject secure connections being attempted if an IPv6 address is provided as the URL.

SlySven committed:
Add: Happy-eyeballs (try IPv4 and IPv6 connections simultaneously)

#### Brief overview of PR changes/additions
Overhauls the profile connection process to identify whether IPv4 and/or
IPv6 protocal addresses are available for the entered details and if BOTH
are to then try to connect with both at the same time (with a slight
preference for IPv6 {by starting it first}). The first one to connect
(including securely and/or via a proxy) wins out and terminates the
process using the other address protocol. This technique is known as
[Happy Eyeballs](https://en.wikipedia.org/wiki/Happy_Eyeballs).

#### Motivation for adding to Mudlet
To improve the end-user's experince if they are fortunate to be on a
"dual-stack" (both protocols available) network.

#### Other info (issues closed, discussion etc)
The on-screen messages have been massaged somewhat to inform the user
about the nature of the connection(s) being made. ALL the addresses that
have been found for the Server are listed instead of just the first one
{this includes a "numerus" translatable string which will need processing
by yours truely!} Should a raw IP address be provided instead of a "URL"
then the result of the "reverse" lookup (trying to get a host name from
the IP address) is displayed instead of just repeating the address - if
anything *is* found - if not that is also noted instead.

One other effect of Mudlet being given a raw IP address is that those are
not normally included in an SSL/TLS certificate - so rendering a secure
connetion impossible in that case - this was already noted
for an IPv4 address but NOT for an IPv6 one. This PR corrects that to
also reject secure connections being attempted if an IPv6 address is
provided as the URL.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
I've set this as ***Draft*** as I want others to test it for themselves as I can't test all aspects of it. I think it does what it says it does but feedback from others will help to check it is feature complete - as it is I cannot test all aspects myself as my ISP only does IPv4 for starters! I need feedback from users of "proxies" and those who have experienced problems with "broken" IPv6 connectivity to a MUD.

SlySven committed:
Fix: appease CodeFactor

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented on src/ctelnet.cpp:
```diff
@@ -299,23 +300,36 @@ void cTelnet::encodingChanged(const QByteArray& requestedEncoding)
 
-QAbstractSocket::SocketError cTelnet::error()
+std::optional<QAbstractSocket::SocketError> cTelnet::error() const
 {
-    return mpSocket.error();
+    if (!mpSocket) {
+        return {};
```
Nice code style with the optional return value and a `{}` 🙂

SlySven replied:
I had to go for a `std::optional<T>` as there didn't seem to be a "No error" value...! :raised_eyebrow:

vadi2 commented on src/ctelnet.cpp:
```diff
@@ -544,82 +675,130 @@ void cTelnet::slot_socketDisconnected()
+            if (sslerr) {
+                // We also have SSL/TLS errors that are not ones we are ignoring
+                // so do not auto-reconnect:
+                mDontReconnect = true;
 
-        if (sslerr) {
-            mDontReconnect = true;
+                for (const auto& error : sslErrors) {
+                    reasons.append(error.errorString());
+                }
+            }
 
-            for (int a = 0; a < sslErrors.count(); ++a) {
-                reason.append(qsl("        %1\n").arg(QString(sslErrors.at(a).errorString())));
+            if (reasons.count()) {
+                postMessage(tr("[ ALERT ] - Socket got disconnected, for %n reason(s):\n"
+                               "%1",
+                               // Intentional comment to separate arguments
+                               "",
+                               reasons.count())
+                                    .arg(reasons.join(QChar::LineFeed)));
+            } else {
```
<img width="583" height="44" alt="Image" src="https://github.com/user-attachments/assets/493c9772-ba39-4df5-b164-6cbd0eb0b543" />

This looks a bit awkward, what do you think about making a friendlier exception

```suggestion

            if (reasons.count() > 1) {
                postMessage(tr("[ ALERT ] - Socket got disconnected, for %n reason(s):\n"
                               "%1",
                               // Intentional comment to separate arguments
                               "",
                               reasons.count())
                                    .arg(reasons.join(QChar::LineFeed)));
            } else if (reasons.count() == 1) {
                postMessage(tr("[ ALERT ] - Socket got disconnected: "
                               "%1",
                               // Intentional comment to separate arguments
                               "")
                                    .arg(reasons.at(1));
            } else {
```

SlySven replied:
:earth_africa: **I18n Not necessary!**

The `en` locales can be translated so that the `%n` value being `1` can indeed not mention the number of reasons. `linguist` will moan (a little) about not having an %n in the singular case but I'm not even sure that CrowdIn will object...

Other locales with different pluralisation rules can be different so extra C++ coding for a singular case is actually an anti-pattern IMVHO...

BTW the only reason for the `// Intentional comment...` and the `""` is because of the need to provide the third numeric arguement to the `QObject::tr(...)` method.

Harrison-Teeg replied:
The nittiest of nitpicks, but it can just be "reasons" now instead of "reason(s)"

SlySven replied:
That is necessary in the "Engineering English" (source code) form but will be handled by the translation system when I ***don't*** make the change Vadi suggests.

Harrison-Teeg replied:
Aha, I get it now. My mistake!

Kebap replied:
Agree with Vadi here. 
The english translation for (1) will still be like "Socket got disconnected, for a reason:\n" and this reads clunky.
How about "Socket got disconnected. Reason(s) is/are:\n" and don't even mention the number in the text.

SlySven replied:
What, and avoid counting the number of fails? :raised_eyebrow::smile:

It is down to what fits for each locale - when there are multiple things it can be helpful to say how many there are - at least the `en_US` AND `en_GB` translations could just as easily be something of the form:
  `[ ALERT ] -  Socket got disconnected, because:\n`
  `             There is a problem between keyboard and chair.`
               
Or
  `[ ALERT ] -  Socket got disconnected, for 3 reasons:\n`
  `             The certificate was self-signed\n`
  `             The certificate expired on 2025/04/01\n`
  `             There is a problem between keyboard and chair.`

vadi2 replied:
Okay.

vadi2 commented on src/ctelnet.cpp:
```diff
@@ -414,59 +428,116 @@ void cTelnet::connectIt(const QString& address, int port)
+     * wish to use the wide versions of these character which can be used here.
+     */
+    const QString displayAddress = isRawIPv6Address(mHostUrl) ? tr("[%1]").arg(mHostUrl) : mHostUrl;
+    postMessage(tr("[ INFO ]  - Looking up the details of server: %1:%2 ...").arg(displayAddress, QString::number(port)));
```
Nice, clear messages 👍

vadi2 reviewed (commented):
Looks perfect to me really. Great job with this! I tested MUDs that support IPv4 and 6 and things seem to work well:

<img width="869" height="176" alt="Image" src="https://github.com/user-attachments/assets/959472be-17d9-4c25-8d07-8dfce3638386" />

AbigailBrady commented:
i'm very pleased to see this. It should be noted that 

> To improve the end-user's experience if they are fortunate to be on a "dual-stack" (both protocols available) network.

it's even better than that! I had one of my users not able to log in to our new site, because we had an AAAA address available and his computer *wasn't* on a dual-stack network, but it thought it did, so the client would just sit there trying to open a connection to a nonroutable network. so it improves end user experience for potentially everyone.

Kebap commented on src/ctelnet.cpp:
```diff
@@ -414,59 +428,116 @@ void cTelnet::connectIt(const QString& address, int port)
+     * separate the latter from the former, however some Far-East locales may
+     * wish to use the wide versions of these character which can be used here.
+     */
+    const QString displayAddress = isRawIPv6Address(mHostUrl) ? tr("[%1]").arg(mHostUrl) : mHostUrl;
+    postMessage(tr("[ INFO ]  - Looking up the details of server: %1:%2 ...").arg(displayAddress, QString::number(port)));
```
```suggestion
    const QString displayAddress = isRawIPv6Address(mHostUrl) ? qsl("[%1]").arg(mHostUrl) : mHostUrl;
    postMessage(tr("[ INFO ]  - Looking up the details of server: %1:%2 ...").arg(displayAddress, QString::number(port)));
```

SlySven replied:
**No!** See lines 473-478 - the comment for translators.

SlySven replied:
Specifically, they might need to be replaced by `［` and `］` U+FF3B and U+FF3D {FULLWIDTH LEFT/RIGHT SQUARE BRACKET} - especially if a numeric IP address is written with `０` to `９` U+FF10 to U+FF19 {FULLWIDTH DIGIT ZERO to NINE}.

Kebap replied:
Did that ever happen though or was requested? Reeks of premature optimization to me.

SlySven replied:
To me the opposite smells of Western-centric mindset/thinking; i.e. thinking all users are just using ASCII characters/encoding.

Kebap commented on src/ctelnet.cpp:
```diff
@@ -414,59 +428,116 @@ void cTelnet::connectIt(const QString& address, int port)
+     * separate the latter from the former, however some Far-East locales may
+     * wish to use the wide versions of these character which can be used here.
+     */
+    const QString displayAddress = isRawIPv6Address(mHostUrl) ? tr("[%1]").arg(mHostUrl) : mHostUrl;
+    postMessage(tr("[ INFO ]  - Looking up the details of server: %1:%2 ...").arg(displayAddress, QString::number(port)));
```
```suggestion
    const QString displayAddress = isRawIPv6Address(mHostUrl) ? tr("[%1]").arg(mHostUrl) : mHostUrl;
    //: %1 is the server address, %2 the port number
    postMessage(tr("[ INFO ]  - Looking up the details of server: %1:%2 ...").arg(displayAddress, QString::number(port)));
```

SlySven replied:
:+1: Yes. a translator comment is a good idea - although I will make it say:
`//: %1 is the server URL or IP address, %2 the port number`

Kebap commented on src/ctelnet.cpp:
```diff
@@ -484,28 +555,86 @@ void cTelnet::slot_send_pass()
+        }
+    } else {
+        if (&mSocket_ipV6 == mpSocket) {
+            postMessage(tr("[  OK  ]  - An open IPv6 connection has been established successfully."));
+        } else {
+            postMessage(tr("[  OK  ]  - An open IPv4 connection has been established successfully."));
```
Please change "open" to "plaintext" or "insecure" etc.
Because "open" is too ambigous, especially when talking about establishing (opening?) a connection.

I think "successfully" can be dropped from all the phrases in this region. It adds no value to the sentence. Can we even establish a connection unsuccessfully?

SlySven replied:
> ... Can we even establish a connection unsuccessfully?

Well, trying to connect to "The Last Outpost" (which does have both types of connections and supports both IP versions) rapidly disconnects you if you do try to connect its secure port in an open manner. This counts as a connection in my book - but not for long... :smiling_imp:

The trouble with saying "insecure" is that it casts negative aspersions towards a Game Server that might not offer a "secure" one. I could agree to `plaintext` but is there any alternatives anyone else has that does not speak so strongly of "encryption"? To me, "plaintext" seems to be generally used to describe a message before encryption or after decryption - and in the "open" case there isn't any of that going on!

Note that if Mudlet is compiled without SSL/TSL support I do not mention "secure"/"open" at all!

Kebap replied:
I get those aspersions, too - and browsers have been casting them on http webservers for years now until eventually Let's Encrypt etc made it easy enough to "encrypt everything" - and not see an insecure broken lock symbol anymore.

<img width="661" height="377" alt="grafik" src="https://github.com/user-attachments/assets/e6eac83f-5288-4f31-abaa-a176c9c3630a" />

Anyway, maybe that's all out of scope here, and we move discussing these text to an issue of its own, while we appreciate your efforts with the happy eyeballs itself. On the other hand, I don't want to give a round of texts to translators and then change them afterwards again. So I felt like we should strive for good texts the first time around. Meh.

By the way, still a draft?

vadi2 replied:
I think an "open connection" is an OK term, should anyone get wildly confused by what this means they can let us know. Agree that "successfully" is redundant, but at the same time this text has already been in use for decades, hasn't been an issue either so not sure why would it be one now.

We should aim to minimize work for translators for sure,  but perhaps let's aim to look ar the big things that matter - such as testing in various network conditions.

Kebap commented on src/ctelnet.cpp:
```diff
@@ -630,38 +809,219 @@ void cTelnet::slot_socketSslError(const QList<QSslError>& errors)
+    for (const auto& address : addressList_ipV6) {
+        /*: Used to add an IPv6 address line to the list displayed during
+         * connecting to a Host.
+         */
+        addressesToReport << tr("%1 (IPv6)").arg(address);
```
What do you want translators to translate here?
Isn't IPv6 an established technical term?
Not sure if this can be qsl() instead.

SlySven replied:
Well, Google translate converts `(IPv6)` into `（IPv6）` for the "Chinese (Simplified)" locale (with full width parentheses) - and that locale probably *doesn't* include a space between it and the preceding IP address...

Kebap commented on src/ctelnet.cpp:
```diff
@@ -630,38 +809,219 @@ void cTelnet::slot_socketSslError(const QList<QSslError>& errors)
+            connect(&mSocket_ipV6, &QIODevice::readyRead, this, &cTelnet::slot_socketReadyToBeRead, Qt::UniqueConnection);
+            connect(&mSocket_ipV4, &QIODevice::readyRead, this, &cTelnet::slot_socketReadyToBeRead, Qt::UniqueConnection);
+
+            if (mConnectViaProxy) {
+                postMessage(tr("[ INFO ]  - Trying secure (IPv4 and IPv6) connections to %1:%2 via proxy...")
```
```suggestion
            if (mConnectViaProxy) {
                //: %1 is the server address, %2 is the port number
                postMessage(tr("[ INFO ]  - Trying secure (IPv4 and IPv6) connections to %1:%2 via proxy...")
```

SlySven replied:
Yeah, probably worth doing though %1 may be a URL or an IP address - if the latter is an IPv6 one then it'll be wrapped in those square brackets...

Kebap commented on src/ctelnet.cpp:
```diff
@@ -630,38 +809,219 @@ void cTelnet::slot_socketSslError(const QList<QSslError>& errors)
+            if (mConnectViaProxy) {
+                postMessage(tr("[ INFO ]  - Trying secure (IPv4 and IPv6) connections to %1:%2 via proxy...")
+                                    .arg(hostInfo.hostName(), QString::number(mHostPort)));
+            } else {
+                postMessage(tr("[ INFO ]  - Trying secure (IPv4 and IPv6) connections to %1:%2 ...")
```
```suggestion
            } else {
                //: %1 is the server address, %2 is the port number
                postMessage(tr("[ INFO ]  - Trying secure (IPv4 and IPv6) connections to %1:%2 ...")
```

Kebap commented on src/ctelnet.cpp:
```diff
@@ -630,38 +809,219 @@ void cTelnet::slot_socketSslError(const QList<QSslError>& errors)
+                connect(&mSocket_ipV6, &QAbstractSocket::disconnected, this, &cTelnet::slot_socketDisconnected, Qt::UniqueConnection);
+                connect(&mSocket_ipV6, &QIODevice::readyRead, this, &cTelnet::slot_socketReadyToBeRead, Qt::UniqueConnection);
+
+                if (mConnectViaProxy) {
+                    postMessage(tr("[ INFO ]  - Trying secure (IPv6) connection to %1:%2 via proxy...")
```
```suggestion
                if (mConnectViaProxy) {
                    //: %1 is the server address, %2 is the port number
                    postMessage(tr("[ INFO ]  - Trying secure (IPv6) connection to %1:%2 via proxy...")
```

Kebap commented on src/ctelnet.cpp:
```diff
@@ -630,38 +809,219 @@ void cTelnet::slot_socketSslError(const QList<QSslError>& errors)
+                    postMessage(tr("[ INFO ]  - Trying secure (IPv6) connection to %1:%2 via proxy...")
+                                        .arg(hostInfo.hostName(), QString::number(mHostPort)));
+                } else {
+                    postMessage(tr("[ INFO ]  - Trying secure (IPv6) connection to %1:%2 ...")
```
```suggestion
                } else {
                    //: %1 is the server address, %2 is the port number
                    postMessage(tr("[ INFO ]  - Trying secure (IPv6) connection to %1:%2 ...")
```

Kebap commented on src/ctelnet.cpp:
```diff
@@ -630,38 +809,219 @@ void cTelnet::slot_socketSslError(const QList<QSslError>& errors)
+                connect(&mSocket_ipV4, &QAbstractSocket::disconnected, this, &cTelnet::slot_socketDisconnected, Qt::UniqueConnection);
+                connect(&mSocket_ipV4, &QIODevice::readyRead, this, &cTelnet::slot_socketReadyToBeRead, Qt::UniqueConnection);
+
+                if (mConnectViaProxy) {
+                    postMessage(tr("[ INFO ]  - Trying secure (IPv4) connection to %1:%2 via proxy...")
```
```suggestion
                if (mConnectViaProxy) {
                    //: %1 is the server address, %2 is the port number
                    postMessage(tr("[ INFO ]  - Trying secure (IPv4) connection to %1:%2 via proxy...")
```

Kebap commented on src/ctelnet.cpp:
```diff
@@ -630,38 +809,219 @@ void cTelnet::slot_socketSslError(const QList<QSslError>& errors)
+                if (mConnectViaProxy) {
+                    postMessage(tr("[ INFO ]  - Trying secure (IPv4) connection to %1:%2 via proxy...")
+                                        .arg(hostInfo.hostName(), QString::number(mHostPort)));
+                } else {
+                    postMessage(tr("[ INFO ]  - Trying secure (IPv4) connection to %1:%2 ...")
```
```suggestion
                } else {
                    //: %1 is the server address, %2 is the port number
                    postMessage(tr("[ INFO ]  - Trying secure (IPv4) connection to %1:%2 ...")
```

Kebap commented on src/ctelnet.cpp:
```diff
@@ -630,38 +809,219 @@ void cTelnet::slot_socketSslError(const QList<QSslError>& errors)
+            connect(&mSocket_ipV6, &QIODevice::readyRead, this, &cTelnet::slot_socketReadyToBeRead, Qt::UniqueConnection);
+            connect(&mSocket_ipV4, &QIODevice::readyRead, this, &cTelnet::slot_socketReadyToBeRead, Qt::UniqueConnection);
+
+            if (mConnectViaProxy) {
+                postMessage(tr("[ INFO ]  - Trying open (IPv4 and IPv6) connections to %1:%2 via proxy...")
```
```suggestion
            if (mConnectViaProxy) {
                    //: %1 is the server address, %2 is the port number
                postMessage(tr("[ INFO ]  - Trying open (IPv4 and IPv6) connections to %1:%2 via proxy...")
```

Kebap replied:
Again, "open" is problematic in these sentences, please adjust

Kebap commented on src/ctelnet.cpp:
```diff
@@ -630,38 +809,219 @@ void cTelnet::slot_socketSslError(const QList<QSslError>& errors)
+                                    .arg(hostInfo.hostName(), QString::number(mHostPort)));
             } else {
-                postMessage(qsl("%1\n").arg(tr("[ INFO ]  - Trying to connect to %1:%2 via proxy...").arg(mHostAddress.toString(), QString::number(mHostPort))));
+                postMessage(tr("[ INFO ]  - Trying open (IPv4 and IPv6) connections to %1:%2 ...")
```
```suggestion
                //: %1 is the server address, %2 is the port number
                postMessage(tr("[ INFO ]  - Trying open (IPv4 and IPv6) connections to %1:%2 ...")
```

Kebap commented on src/ctelnet.cpp:
```diff
@@ -630,38 +809,219 @@ void cTelnet::slot_socketSslError(const QList<QSslError>& errors)
+                connect(&mSocket_ipV6, &QIODevice::readyRead, this, &cTelnet::slot_socketReadyToBeRead, Qt::UniqueConnection);
+
+                if (mConnectViaProxy) {
+                    postMessage(tr("[ INFO ]  - Trying open (IPv6) connection to %1:%2 via proxy...")
```
```suggestion
                    //: %1 is the server address, %2 is the port number
                    postMessage(tr("[ INFO ]  - Trying open (IPv6) connection to %1:%2 via proxy...")
```

Kebap commented on src/ctelnet.cpp:
```diff
@@ -630,38 +809,219 @@ void cTelnet::slot_socketSslError(const QList<QSslError>& errors)
+                    postMessage(tr("[ INFO ]  - Trying open (IPv6) connection to %1:%2 via proxy...")
+                                        .arg(hostInfo.hostName(), QString::number(mHostPort)));
+                } else {
+                    postMessage(tr("[ INFO ]  - Trying open (IPv6) connection to %1:%2 ...")
```
```suggestion
                } else {
                    //: %1 is the server address, %2 is the port number
                    postMessage(tr("[ INFO ]  - Trying open (IPv6) connection to %1:%2 ...")
```

Kebap commented on src/ctelnet.cpp:
```diff
@@ -630,38 +809,219 @@ void cTelnet::slot_socketSslError(const QList<QSslError>& errors)
+                connect(&mSocket_ipV4, &QAbstractSocket::disconnected, this, &cTelnet::slot_socketDisconnected, Qt::UniqueConnection);
+                connect(&mSocket_ipV4, &QIODevice::readyRead, this, &cTelnet::slot_socketReadyToBeRead, Qt::UniqueConnection);
+
+                if (mConnectViaProxy) {
+                    postMessage(tr("[ INFO ]  - Trying open (IPv4) connection to %1:%2 via proxy...")
```
```suggestion
                if (mConnectViaProxy) {
                    //: %1 is the server address, %2 is the port number
                    postMessage(tr("[ INFO ]  - Trying open (IPv4) connection to %1:%2 via proxy...")
```

Kebap commented on src/ctelnet.cpp:
```diff
@@ -630,38 +809,219 @@ void cTelnet::slot_socketSslError(const QList<QSslError>& errors)
+                    postMessage(tr("[ INFO ]  - Trying open (IPv4) connection to %1:%2 via proxy...")
+                                        .arg(hostInfo.hostName(), QString::number(mHostPort)));
+                } else {
+                    postMessage(tr("[ INFO ]  - Trying open (IPv4) connection to %1:%2 ...")
```
```suggestion
                    //: %1 is the server address, %2 is the port number
                    postMessage(tr("[ INFO ]  - Trying open (IPv4) connection to %1:%2 ...")
```

Kebap commented on src/ctelnet.cpp:
```diff
@@ -3271,58 +3631,70 @@ void cTelnet::setMSPVariables(const QByteArray& msg)
-        msgBox->setInformativeText(tr("Update to port %1 and connect with encryption?").arg(QString::number(mpHost->mMSSPTlsPort)));
-        msgBox->setStandardButtons(QMessageBox::Yes | QMessageBox::No);
-        msgBox->setDefaultButton(QMessageBox::Yes);
+        postMessage(tr("[ INFO ]  - A more secure connection on port %1 is available.").arg(QString::number(mpHost->mMSSPTlsPort)));
```
How is this now "more" secure?
Earlier in the code we called them "secure" connections vs "open" connections, right?

edit: Sorry, it seems like this phrase was not altered in this commit at all. Still bogus.

mpconley replied:
The phrase "more secure" is used in cybersecurity and other fields to indicate a higher level of protection against unauthorized access, threats, or vulnerabilities. For example, a passphrase is considered more secure than a password because it's longer, easier to remember, and thus harder for cybercriminals to guess or crack. 

What's going on @Kebap ? You're going into dark places lately and turning over everyone elses work.

vadi2 replied:
I agree it is not that important to nitpick wording like this, but at the same time let's remain civil.

Kebap commented on src/dlgConnectionProfiles.cpp:
```diff
@@ -1875,26 +1875,25 @@ bool dlgConnectionProfiles::validateProfile()
+             */
+            notificationAreaMessageBox->setText(qsl("%1%2\n\n%3").arg(!notificationAreaMessageBox->text().isEmpty() ? notificationAreaMessageBox->text().append(QChar::LineFeed) : QString(),
+                                                                      tr("Please enter the URL of the Game server.\n\n"
+                                                                         "<i>SSL connections require the URL of the Game server.</i>"),
```
This changed text now seems pretty redundant.

Also SSL is deprecated, TLS the new way to go.
Anyway, too much technical detail for players.

```suggestion
                                                                      tr("For secure connections, please enter the URL of the Game server."),
```

SlySven replied:
The "certificate" used for the encryption is extremely unlikely to include the IP Address as an identifier for the Server - indeed we did reject any "raw" (quad dot separated numbers each between 0 and 255) IPv4 addresses if that was previously supplied as the "URL" for a "secure" connection in the past (this PR extends that to also reject raw IPv6 ones in the same way) - anyhow this is the warning message for the user so this explanation (without the technical reasons behind it) why Mudlet is rejecting what was entered seems reasonable IMHO. 

I'd consider changing `SSL` to `SSL/TLS` but I think it sounds/looks better to say that rather than reusing "secure" yet again - besides even though SSL is deprecated that doesn't stop it from being used!

SlySven commented on src/ctelnet.cpp:
```diff
@@ -1019,13 +1379,13 @@ QString cTelnet::decodeOption(const unsigned char ch) const
+    const bool connected = (mpSocket && mpSocket->state() == QAbstractSocket::ConnectedState);
 
-    if (mHostName.isEmpty() && mHostPort == 0) {
+    if (mHostUrl.isEmpty() || mHostPort == 0) {
```
TBH I'm not sure if these tests should be an AND or an OR - thoughts anyone?

Kebap commented:
Here's a game server you can test:
- Name:    mg.mud.de
- Addresses:  
  - 2a03:4000:5f:56:4f9:5aff:fe68:f69
  - 89.58.11.82

Or find current Addresses with `nslookup` yourself

SlySven committed:
Revise: provide information to translators for new texts in PR.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: warning message about using an IP address for a secure connection

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Add_HappyEyeballs

SlySven commented:
@mpconley Does the stuff I think you originally did to inform users that they could update their "open" connection to a "secure" one by changing the port number via an OOB method (IIRC) still work after this PR. There is also some code that reports on types of "secure" connection errors associated with "iffy" certificates that the end-user can "override"/"bypass" if they want to - does it still work here - or have I messed it up? Again I think you did the original work and presumably know how to check/confirm it works. If that is the case can you test out this PR - and if it is not, then please let me know so I can try and track down someone else who can?

mpconley commented:
> Does the stuff I think you originally did to inform users that they could update their "open" connection to a "secure" one by changing the port number via an OOB method (IIRC) still work after this PR.

Works great. Just try port stickmud.com 7680 and you get upgraded to 7670 along the way. I closed Mudlet, reopened, and verified the connection is 7670 (encrypted). Also the GMCPAuthenticator code works, too.

https://github.com/user-attachments/assets/601f068c-c9f5-4e84-9c4d-a2a7d2787765

mpconley commented:
Two more things:

There is a double connection message in that video I noticed and I reproduced it again after wiping the profile and trying again. I've seen that before elsewhere. I've fixed it a PR I have not yet released. I don't recall the fix offhand, but I could look this weekend or help you find it with LLM powers.

Another one is that I experienced a 15-20 second long delay for the profile to come up and start connecting after the profile wipe and remaking it. It worked as expected but there was that long delay.

<img width="2190" height="1548" alt="Screenshot 2025-09-04 at 8 27 01 PM" src="https://github.com/user-attachments/assets/fff5b2e5-15ac-4ae2-a1d5-73cd559c9bf6" />

https://github.com/user-attachments/assets/e1f60c46-0942-4b47-81e9-1dc59e4d79b5

Happy to help with more testing this weekend.

Kebap commented:
Given #8209 heavily reduces the connection details shown to mere players, this PR #8135 here needs some more thought.

See discussion there for more details, also more testing that seems necessary here (not sure if now or after draft mode ends)

My humble suggestion would include moving all the gory connection details to the Mudlet debug console for the interested few.

They could then open a profile in offline mode, open the debug console, and only then start connection, to see details unfold.

It mimics the way modern browsers hide the connection details like IP, port, protocol, etc from the mere surfers of the webz.

SlySven committed:
Merge branch 'development' into Add_HappyEyeballs

SlySven committed:
Revise: incorporate proposal of #8209 into this PR

Move some of the "gory" detail about the connection process to the Central
Debug Console from the main console for each profile - as #8209 proposed to
do so.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: clean up some LineFeed issues with the revised messages

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Add_HappyEyeballs

Conflicts resolved in:
* src/ctelnet.cpp
* src/dlgProfilePreferences.cpp

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
I've confirmed to myself that the display of the certificate details in the "Connection" tab of the preferences still works so the only untested bit is, I think, using a SOCKS proxy for a connection. This was introduced by @vadi2 so presumable he knows how to test it. As such I'll mark this as now ready for review but not merge it until he gives a go-ahead...

vadi2 reviewed (approved):
Still works - 
<img width="1818" height="1040" alt="image" src="https://github.com/user-attachments/assets/d56be774-6111-4ade-b5b5-b73993d73ed8" />

vadi2 commented:
Lets merge this?

SlySven committed:
Merge branch 'development' into Add_HappyEyeballs

Conflicts resolved in:
* src/CMakeLists.txt
* src/ctelnet.cpp

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>
